trusts/decorators.py

```diff
@@ -119,4 +194,31 @@ def check_perms(request, *args, **kwargs):
             raise PermissionDenied
         return False
 
+    def check_by_p(request, *args, **kwargs):
+        perms = perm.get_leaves()
+
+        for p in perms:
+            fieldlookups = _resolve_fieldlookups(request, kwargs, **p.get_lookup_kwargs())
+            items = None
+            if fieldlookups is not None:
+                items = _get_permissible_items(request, p.perm, fieldlookups)
+                if items is None:
+                    p.permitted = False
```

Comment:
The only problem here it's if you want to use P class like this:
`
p1 = P('app.moderate_xyz', pk=K('xyz_id')) 
p2 = P('app.update_xyz:own', pk=K('xyz_id'))
p3 = P('app.delete_xyz:own', pk=K('xyz_id'))

@permission_required(p1 | p2)
def edit_xyz_view(request, xyz_id):
    pass

@permission_required(p2 & p3)
def edit_xyz_view(request, xyz_id):
    pass
`

But it's easy to solve, we just have to use copy() or deepcopy() in __and__ and __or__ methods.